### PR TITLE
feature/fingerbank_api

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Fingerbank API

Delete branch after merge
=========================
NO

Checklist
=========
- [ ] Document the feature
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

**Comment by extrafu** [comment]
@julsemaan Please review/comment ASAP in order to get it in.

File: lib/pf/UnifiedApi/Controller/Fingerbank.pm [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+sub remove {
+    my ($self) = @_;
+    my $id = $self->id;
+    my ($status, $msg) = $self->fingerbank_model->delete($id);
+    if (is_error($status)) {
+        return $self->render_error($status, $msg);
+    }
+
+    return $self->render(json => {});;
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Fingerbank.pm:131]: Not sure its a plain 200 the other deletion calls return

File: lib/pf/UnifiedApi/Controller/Fingerbank.pm [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    my ($self) = @_;
+    url_unescape($self->stash->{$self->url_param_name});
+}
+
+sub build_list_search_info {
+    my ($self) = @_;
+    my $params = $self->req->query_params->to_hash;
+    my $scope = $self->stash->{scope};
+    my $model = $self->fingerbank_model;
+    my $db = fingerbank::DB_Factory->instantiate(schema => 'Local');
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Fingerbank.pm:150]: We can't search in the upstream schema or is this just for metadata ?

File: lib/pf/UnifiedApi/Controller/Fingerbank/UserAgent.pm [hunk 3]
```diff
@@ -0,0 +1,48 @@
+package pf::UnifiedApi::Controller::Fingerbank::UserAgent;
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Controller/Fingerbank/UserAgent.pm]: Why duplicate controllers for files that contain exactly the same thing ?

Anything different here than from the other resources ?

File: lib/pf/UnifiedApi/Search/Builder/Fingerbank.pm [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+    my ($self, $search_info) = @_;
+    my ($status, $search_args) = $self->make_search_args($search_info);
+    if ( is_error($status) ) {
+        return $status, $search_args;
+    }
+
+    my $model = $search_info->{model};
+    my $limit = $search_args->{'-limit'};
+    my $offset = $search_args->{'-offset'};
+    #use Data::Dumper;print Dumper($search_args);
```
Comment (julsemaan) [inline lib/pf/UnifiedApi/Search/Builder/Fingerbank.pm]: remove

File: t/unittest/UnifiedApi/Plugin/RestCrud.t [hunk 5]
```diff
@@ -1,490 +0,0 @@
-#!/usr/bin/perl
```
Comment (julsemaan) [inline t/unittest/UnifiedApi/Plugin/RestCrud.t:1]: Looks like this is gone but not replaced by anything else.

Anyway we can keep the same level of testing ?

**Review by julsemaan — CHANGES_REQUESTED** [review]

**Comment by julsemaan** [comment]
We're also missing the spec for the new API endpoints

File: lib/pf/UnifiedApi/Controller/Fingerbank.pm (hunk 1 above)
Comment (jrouzierinverse) [reply lib/pf/UnifiedApi/Controller/Fingerbank.pm:131]: It is

File: lib/pf/UnifiedApi/Controller/Fingerbank.pm (hunk 2 above)
Comment (jrouzierinverse) [reply lib/pf/UnifiedApi/Controller/Fingerbank.pm:150]: This is for metadata

File: lib/pf/UnifiedApi/Controller/Fingerbank.pm (hunk 1 above)
Comment (jrouzierinverse) [reply lib/pf/UnifiedApi/Controller/Fingerbank.pm:131]: It is

File: lib/pf/UnifiedApi/Controller/Fingerbank.pm (hunk 2 above)
Comment (jrouzierinverse) [reply lib/pf/UnifiedApi/Controller/Fingerbank.pm:150]: Just for meta

File: lib/pf/UnifiedApi/Controller/Fingerbank/UserAgent.pm (hunk 3 above)
Comment (jrouzierinverse) [reply lib/pf/UnifiedApi/Controller/Fingerbank/UserAgent.pm]: This is a mistake will fix

File: t/unittest/UnifiedApi/Plugin/RestCrud.t (hunk 5 above)
Comment (jrouzierinverse) [reply t/unittest/UnifiedApi/Plugin/RestCrud.t:1]: Same level of testing is not needed.

The plugin only provides very simple shortcuts to the route class.
If the shortcuts did not work then routing would not work.

**Pull request merged**